Document ads library api (#48)
ambanum/disinfo.quaidorsay.fr: _pages/encyclopedia/facebook-ad-library.md

---

> permalink: /encyclopedia/fb-ad-library

Reviewer: ```suggestion
permalink: /encyclopedia/facebook-ads-library
```

> nav_order: 10

Reviewer: We'll rather not show this in the navigation for the moment.
```suggestion
nav_exclude: true
```

---

> # Facebook Ad Library

Reviewer: The official name is apparently Ads Library.

```suggestion
# Facebook Ads Library
```

## Limitations

### Reports are sketchy

> Facebook provides reports for some countries including France. The [report covering France](https://www.facebook.com/ads/library/report/?source=archive-landing-page&country=FR) counts 15.681 ads.

Reviewer: As of when?

> Facebook provides reports for some countries including France. The [report covering France](https://www.facebook.com/ads/library/report/?source=archive-landing-page&country=FR) counts 15.681 ads.

Reviewer: > As I understand it, US, BR and all EU countries are covered. Might be worth checking for others and be more specific about which countries get a report exactly.

> The data is aggregated by sponsor, and provides no information about the content of the ads.

Reviewer: > Isn't it also limited to the 100 highest payers or something like that?

Author: No I don't think so.

> The CSV file was unavailable on May 27th.

Reviewer: Just for a couple hours, right?

### API and search engine

We [download regularly](https://desinfo.quaidorsay.fr/ads/dumps/) the content of the Ad Library [using its API](https://github.com/ambanum/political-ads-scraper). We observed that some ads get regularly removed from the library, in contradiction with Facebook's pledge to keep the data during seven years.

Date    | Number of ads | New ads | Removed ads
--------|---------------|---------|------------
2019-05-16 11-55-05 | 12838 | NA | NA
2019-05-16 15-30-09 | 12946 | 110 | 2
2019-05-16 20-37-43 | 12934 | 56 | 68
2019-05-20 14-03-03 | 13550 | 666 | 50
2019-05-27 14-53-52 | 9288 | 695 | 4957
2019-05-27 16-23-19 | 9277 | 2 | 13

> 2019-05-27 16-30-54 | 9277 | 0 | 0

Reviewer: > What would be awesome would be to also list the number of ads as shown by Facebook in the report page.

Author: This depends on https://trello.com/c/OCfIa0rU/158-g%C3%A9n%C3%A9rer-un-graphe-du-nombre-de-publicit%C3%A9s-r%C3%A9ellement-disponibles

Author: Now here: https://disinfo.quaidorsay.fr/political-ads/consistency?country=fr

> For example, we can count 13 ads on May 16th and 14 ads on May 20th promoting or criticising François-Xavier Bellamy's campain for the EU elections. On May 27th, only 2 of those ads are still visible through both the search engine and the API.

Reviewer: ```suggestion
For example, we can count 13 ads on May 16th and 14 ads on May 20th promoting or criticizing a French MEP candidate for the EU elections. On May 27th, only 2 of these ads are still visible through both the search engine and the API.
```

Autour de François Xavier Bellamy, d’Agnès Evren et d’Arnaud Danjean, notre famille politique présentent une liste de conviction composée de candidats dévoués, compétents et utiles pour défendre nos intérêts et porter haut et fort nos convictions.
Je vous invite dès maintenant à faire campagne et à appeler à voter pour l’avenir de notre pays et de notre continent."

Facebook ID: 2316760961894602
Text: "J'ai ouvert cette page de soutien à Francois Xavier Bellamy parce que je connais ce brillant jeune homme depuis longtemps, et en particulier pour l'avoir croisé et entendu au cours de manifestations ou réunions Alliance Vita (Universités de la Vie) Paray le Monial, soirées philo etc. etc. Son engagement public est un signe d'espérance et un témoignage exemplaire au service du bien commun pour toutes les classes politiques en France. L'engagement civique d'un chrétien, catholique, tel que Francois Xavier Bellamy, apportera au débat public une vision du monde, de la France et de l'Europe mais aussi une pratique et un langage exemplaires dans leur lucidité et leur exigence de vérité et de loyauté."
```

> The API suffers from frequent server-side errors. For example, since May 16th, 2019 we have not been able to downloads ads shown in Germany.

Reviewer: >  Can we be more specific about the kind of server-side errors we get?

Author: Notes to myself:

*400, '{"error":{"message":"(#2) Service temporarily unavailable","type":"OAuthException","is_transient":true,"code":2,"fbtrace_id":"HhJYNJSSOMX"}}'

*500, '{"error":{"code":1,"message":"An unknown error occurred","error_subcode":99}}

*500, '{"error":{"code":1,"message":"Please reduce the amount of data you\'re asking for, then retry your request"}} (when the page size is below the documented value of 5000)

*And yet another error with an HTML page containing the message "Sorry, we are working on this" but I lost the logs.

Author: Just got one!

400
```
<!DOCTYPE html>
<html lang="en" id="facebook">
  <head>
    <title>Facebook | Error</title>
    <meta charset="utf-8">
    <meta http-equiv="cache-control" content="no-cache">
    <meta http-equiv="cache-control" content="no-store">
    <meta http-equiv="cache-control" content="max-age=0">
    <meta http-equiv="expires" content="-1">
    <meta http-equiv="pragma" content="no-cache">
    <meta name="robots" content="noindex,nofollow">
    <style>
      html, body {
        color: #141823;
        background-color: #e9eaed;
        font-family: Helvetica, Lucida Grande, Arial,
                     Tahoma, Verdana, sans-serif;
        margin: 0;
        padding: 0;
        text-align: center;
      }

      #header {
        height: 30px;
        padding-bottom: 10px;
        padding-top: 10px;
        text-align: center;
      }

      #icon {
width: 30px;
      }

      h1 {
        font-size: 18px;
      }

      p {
        font-size: 13px;
      }

      #footer {
        border-top: 1px solid #ddd;
 color: #9197a3;
        font-size: 12px;
        padding: 5px 8px 6px 0;
      }
    </style>
  </head>
  <body>
    <div id="header">
      <a href="//www.facebook.com/">
       <img id="icon" src="//static.facebook.com/images/logos/facebook_2x.png" />
      </a>
    </div>
    <div id="core">
      <h1 id="sorry">Sorry, something went wrong.</h1>
      <p id="promise">
        We're working on it and we'll get it fixed as soon as we can.
      </p>
      <p id="back-link">
        <a id="back" href="//www.facebook.com/">Go Back</a>
      </p>
      <div id="footer">
        Facebook
        <span id="copyright">
          &copy; 2019
        </span>
        <span id="help-link">
          &#183;
          <a id="help" href="//www.facebook.com/help/">Help Center</a>
        </span>
      </div>
    </div>
    <script>
      document.getElementById('back').onclick= function() {
        if (history.length > 1) {
          history.back();
          return false;
        }
      };

      // Adjust the display based on the window size
   if (window.innerHeight < 80 || window.innerWidth < 80) {
        // Blank if window is too small
        document.body.style.display = 'none';
      };
      if (window.innerWidth < 200 || window.innerHeight < 150) {
        document.getElementById('back-link').style.display = 'none';
        document.getElementById('help-link').style.display = 'none';
      };
      if (window.innerWidth < 200) {
        document.getElementById('sorry').style.fontSize = '16px';
      };
      if (window.innerWidth < 150) {
 document.getElementById('promise').style.display = 'none';
      };
      if (window.innerHeight < 150) {
        document.getElementById('sorry').style.margin = '4px 0 0 0';
        document.getElementById('sorry').style.fontSize = '14px';
        document.getElementById('promise').style.display = 'none';
      };
    </script>
  </body>
</html>
```

Author: For twitter ads API : 
* 500, `{"errors":[{"message":"Internal error","code":131}]}`
